### DISCUSSION TITLE
Deduce overloads based on defaults

### DISCUSSION THREAD

**Discussion by Chiggy-Playz** [microsoft/pyright#4234] [Ideas]
Pyright currently does not account the defaults for overloads. It would be nice to have it infer the defaults, preventing repetition of `@overloads`
Example Code:
```py
from typing import overload, Optional, Union, Literal

class X:
    pass

class Y:
    pass

@overload
def f(a: Literal[False], b: Literal[False]) -> Optional[X]:
    ...

@overload
def f(a: Literal[True], b: Literal[False]) -> Optional[Y]:
    ...

@overload
def f(a: Literal[False], b: Literal[True]) -> X:
    ...

@overload
def f(a: Literal[True], b: Literal[True]) -> Y:
    ...

def f(a: bool = False, b: bool = False) -> Union[Optional[X], Optional[Y]]:
    ...

reveal_type(f())
```

Pyright shows an error saying `No overloads for "f" match the provided arguments:  Argument types: ()`
which makes sense, but it would be nice if pyright inferred the defaults of the argument, which would save typing out lots of overloads.

**Comment by erictraut** [comment]
This isn't how overloads work. See PEP 484 for details. When overloads are present, calls are evaluated according to the list of overloaded signatures without regard for the implementation's signature. If you want the overloaded signatures to reflect the presence of defaults, you need to include those defaults in the overload signatures.
